### ISSUE TITLE
A note on numerical preconditions in Softmax gradient translation-invariance testing

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107670]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.3.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.7.12

### CUDA/cuDNN version

cuDNN 8.6

### GPU model and memory

Tesla V100

### Current behavior?

Hello TensorFlow team,

We’d like to report and clarify an observation from metamorphic testing of operator gradients, focusing on how large constant shifts interact with the numerical preconditions of certain metamorphic relations (MRs).

We tested three operators under the same “constant shift” stress condition:

AvgPool2D (AP) 

MaxPool2D (MP)

Softmax

We observed that AvgPool2D consistently satisfies the expected MR, while MaxPool2D and Softmax can exhibit reproducible deviations under large shifts. We have already reported the MaxPool2D case separately; this email summarizes the broader finding and asks for guidance on how best to interpret/document it, especially for Softmax.

Background (the MRs we tested)
1) Constant shift transformation

( X' = X + C )

Where (C) is constant along the MR-relevant axis (e.g., per-window constant for pooling, per-softmax slice constant for Softmax), and may vary across other dimensions via broadcasting.

2) Expected behavior

AvgPool2D:
Forward shifts by a constant; backward should be invariant under the same upstream gradient:
( gX' ≈ gX )

MaxPool2D (with the precondition that argmax routing is preserved):
( gX' ≈ gX )

Softmax (theoretically shift-invariant along the normalization axis):
( \mathrm{softmax}(x + c) = \mathrm{softmax}(x) )
and, under the same upstream gradient (G),
( gX' ≈ gX )

What we observed
AvgPool2D

Under large constant shifts (including magnitudes on the order of 1e4–1e5 in float32), AvgPool2D did not show violations in our metamorphic checks (forward/gradient consistency remained within tolerance).

MaxPool2D

Under settings such as padding=SAME and stride=1, we observe reproducible MR violations with large violation metrics (vR) on both CPU and GPU when (|c|) is sufficiently large (e.g., hundreds or larger).
This MaxPool2D finding has already been reported. https://github.com/tensorflow/tensorflow/issues/107543

Softmax

Under large constant shifts (e.g., 1e4–1e5 in float32) applied as a constant along the softmax axis (via broadcasting), we observe reproducible numerical deviations in:

forward output consistency and/or

backward gradient consistency (gX)

The effect appears implementation-path sensitive (e.g., CPU vs GPU and/or layout paths), though we can provide exact version/platform matrices and minimal repro scripts.

Our analysis / hypothesis

We believe the MaxPool2D behavior is expected under finite precision and ties:

For large (|c|), float32 ULP becomes coarse relative to small differences.

Previously strictly ordered values can become equal or nearly equal after the shift (implicit ties).

This can change argmax routing, hence backward routing changes and the MR precondition (“order preserved”) no longer holds.

For Softmax, we suspect a related but distinct numerical-stability / implementation-path-consistency issue:

Softmax is mathematically shift-invariant along the normalization axis, but under extreme magnitudes in float32, different kernel paths / reduction orders / stabilization behaviors can potentially lead to observable deviations.

The fact that AvgPool2D passes under the same stress condition suggests the phenomenon is not a general issue in the test harness, but operator-specific numerical behavior.

Questions

We’d like to confirm the following points:

For MaxPool2D gradients, is it reasonable to treat violations under large constant shifts as Not Applicable (N/A) when the “argmax routing preserved” precondition is broken by finite-precision tie effects?

For Softmax, is it expected that extremely large constant shifts (float32) may lead to measurable forward/gradient deviations across device/layout implementations, or would you consider this worthy of investigation as an implementation-path consistency / numerical stability issue?

Would it be helpful to add a brief documentation note stating that large constant shifts can affect:

MaxPool gradient routing (due to finite-precision tie effects), and

potentially Softmax numerical consistency in extreme regimes,
even though the underlying mathematical properties hold in real arithmetic?

Reproduction

We have minimal deterministic scripts and concrete configurations (TF version, CPU/GPU, layout) for:

MaxPool2D (already reported), and

Softmax (new observation)

We can provide them in this thread immediately if useful.

Thank you for your time and for the continued work on TensorFlow.

Best regards,
XiaoHuaOvo

### Standalone code to reproduce the issue

```shell
"""
Softmax Gradient Shift-Invariance Bug - Minimal Repro (Gradient Only)
TensorFlow 2.3.0+ / CPU

Theory:
  softmax(x + c) = softmax(x)
  ⇒ dL/dx(x + c) = dL/dx(x)   (for the same upstream gradient G)

Observed:
  When c is very large, gradients show numerical discrepancies.
"""

import numpy as np
import tensorflow as tf
import os

# Force CPU to avoid GPU kernel differences
os.environ['CUDA_VISIBLE_DEVICES'] = ''

print("\n" + "=" * 60)
print("Softmax Gradient Shift-Invariance Bug Reproduction (Gradient Only)")
print("=" * 60)

# 1. Input data
np.random.seed(42)
tf.random.set_seed(42)

x = np.random.uniform(-1, 1, (4, 10, 16, 16)).astype(np.float32)

# 2. Large constant shift
shift = 1_000_000.0
x_shifted = x + shift

print(f"Input range: [{x.min():.2f}, {x.max():.2f}]")
print(f"Shift value: {shift:.0f}")
print(f"Shifted input range: [{x_shifted.min():.0f}, {x_shifted.max():.0f}]")

# 3. Fixed upstream gradient G (same shape as softmax output)
G = tf.random.normal(x.shape, dtype=tf.float32)

# Softmax axis (C dimension in NCHW)
axis = 1

# 4. Compute gradient for original x
x_tf = tf.Variable(x)
with tf.GradientTape() as tape:
    y = tf.nn.softmax(x_tf, axis=axis)
    loss = tf.reduce_sum(y * G)
grad_x = tape.gradient(loss, x_tf).numpy()

# 5. Compute gradient for shifted input x + c
x_shift_tf = tf.Variable(x_shifted)
with tf.GradientTape() as tape:
    y_shift = tf.nn.softmax(x_shift_tf, axis=axis)
    loss_shift = tf.reduce_sum(y_shift * G)
grad_x_shift = tape.gradient(loss_shift, x_shift_tf).numpy()

# 6. Gradient difference analysis
diff = np.abs(grad_x - grad_x_shift)
max_diff = np.max(diff)
mean_diff = np.mean(diff)

print(f"\nMax gradient difference: {max_diff:.6e}")
print(f"Mean gradient difference: {mean_diff:.6e}")

# 7. vR metric (same definition as in the testing framework)
atol, rtol = 1e-5, 1e-3
scale = np.maximum(np.abs(grad_x.ravel()), np.abs(grad_x_shift.ravel()))
tolerance = atol + rtol * scale
r_i = diff.ravel() / (tolerance + 1e-100)

vR = np.max(r_i)
rho = np.sum(r_i > 1.0) / len(r_i)

print(f"\nvR metric: {vR:.2f}")
print(f"Violation detected: {vR > 1.0} (threshold = 1.0)")
print(f"Violation ratio: {rho:.2%}")

if vR > 1.0:
    print("\n⚠️  Gradient shift-invariance numerical violation reproduced!")
    print("   Theory: d/dx softmax(x + c) == d/dx softmax(x)")
    print(f"   Observed: {rho:.1%} of gradient elements exceed tolerance")
else:
    print("\n✓ No violation detected (try a larger shift value)")

print("=" * 60 + "\n")
```

### Relevant log output

```shell
Max gradient difference: 1.138362e-02
Mean gradient difference: 1.049445e-03

vR metric: 211.74
Violation detected: True (threshold = 1.0)
Violation ratio: 95.42%

⚠️  Gradient shift-invariance numerical violation reproduced!
   Theory: d/dx softmax(x + c) == d/dx softmax(x)
   Observed: 95.4% of gradient elements exceed tolerance
============================================================
```

**Comment by akhilgoe** [comment]
Hi @XiaoHuaOvo-tech, I don't think the TF kernel is the problem here. The discrepancy in my opinion arises because the shifted F32 input cannot fully preserve the orignial input, leading to precision loss even before the inputs reach the TF kernel.

Eg:

```
import numpy as np

x = np.float32(np.random.uniform())
shift = 1000000

print("x = ", x)
print("(x + shift - shift) = ", x + shift - shift)
```

Output:
```
x = 0.6594747
(x + shift - shift) = 0.6875
```

If I change the dtype to float64 in your reproducer then the test passes.

**Comment by XiaoHuaOvo-tech** [comment]
Hi @akhilgoe,
﻿
Thanks again for the clear explanation and the reproduction script.
﻿
I completely agree with your assessment. I overlooked the fact that the large shift combined with `float32` precision leads to significant information loss at the input level before the kernel execution, which naturally propagates to the gradient.
﻿
This is indeed a limitation of floating-point arithmetic rather than a logical bug in the operator implementation. I will refine my test case generation to respect the valid numerical domain of `float32`.
